rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,565 @@
+scale.
+
+## Details
+
+### Current State
+
+Currently, the `token` join method introduces a generation counter as a label
+on the Bot user. This counter is contained within the Bot certificate and on
+each renewal, this counter is incremented. When the counter within the certificate
+de-synchronises with the counter on the user, the Bot is locked out as a security
+measure.
+
+The fact that this counter is stored within a label on the Bot user creates a 
+one-to-one binding between a single instance of `tbot` and a single Bot user.
+This is not the case when using the delegated join methods.
+ 
+### Persistent Bot Instance Identity
+
+Today, there is no persistent identifier for an individual instance of a Bot.
+Instead, all `tbot` instances are effectively identified solely by their Bot
+identity. There is no easy way to distinguish them. On each renewal, `tbot`
+regenerates the private-public key pair that is used within its certificate and
+there is no other form of unique ID that is persisted across renewals.
+
+This poses a few challenges:
+
+- For the purposes of auditing, it is not possible to trace actions to a
+  specific instance of a Bot.
+- For improving the `token` join method to support multiple Bot instances
+  associated with a single Bot, there is no identifier to correlate with the 
+  generation counter.
+- For analytics purposes, it's difficult for us to track the number of
+  individual Bot instances in use. We cannot easily determine if it's a single
+  very active Bot instance, or many less active Bot instances.
+
+To rectify this, a unique identifier should be established for an instance of
+a Bot.
+
+#### Public Key Fingerprint
+
+One option is to modify the behaviour of `tbot` to persist and reuse the
+keypair across renewals. We could then use a fingerprint of the public key as a
+unique identifier of the Bot instance.
+
+This feels like a natural identifier. It avoids introducing a new attribute to
+certificates as the public key is already encoded within certificates. The
+nature of public-key cryptography also means that this provides the Bot instance
+a way to identify itself without needing an issued certificate.
+
+To mitigate the risk of pre-image attacks, SHA256 will be used to determine the
+fingerprint of the public key. In addition, the full public key should be
+recorded and verified against when authenticating a Bot instance action.
+
+It should be noted that with this technique, rotating the keypair of a `tbot`
+instance would reset the identity of that instance. Rotation of this keypair
+would be unusual and this side effect seems expected.
+
+##### Alternative: UUID Certificate Attribute
+
+On the initial join of a Bot instance, we could generate a UUID to identify that 
+Bot instance and encode this within the certificate. Upon renewals, the UUID
+would be copied from the current certificate and into the new one.
+
+Whilst this is fairly easy to implement for the `token` join method, one
```

Comment:
This limitation would go away if we moved the mechanism for delegated joins over to this mechanism, correct?